## House Prices - Advanced Regression Techniques

## notes
target = SalePrice  

problems: too many features...

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('./train.csv')

In [3]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [4]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

## Handling the missing values

In [6]:
miss_values = df.isnull().sum()[df.isnull().sum() > 0]
print(miss_values.sort_values(ascending=False))

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
MasVnrType       872
FireplaceQu      690
LotFrontage      259
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
BsmtFinType2      38
BsmtExposure      38
BsmtFinType1      37
BsmtCond          37
BsmtQual          37
MasVnrArea         8
Electrical         1
dtype: int64


In [7]:
# missing values
print(f'{miss_values.sum()} missing values')

# categorical features
cat_features = df.select_dtypes(include=['object']).shape[1]
print(f"{cat_features} categorical features")

# numerical features
num_features = df.select_dtypes(include=['int64', 'float64']).shape[1]
print(f"{num_features} numerical features")

# shape
print(f"\n{df.shape[0]} enteries, {df.shape[1]} features")

7829 missing values
43 categorical features
38 numerical features

1460 enteries, 81 features


## Reports
We have 1460 enteries and 81 features 
7829 missing values  
43 categorical features  
38 numerical features  
whats scary? too many features and too many missing values.

## Training a "baseline" linear model, (( without the missing values ))

In [6]:
# Select only numerical columns
df_numeric = df.select_dtypes(include=['int64', 'float64'])

# Drop columns with any missing values 
df_cl_baseline = df_numeric.dropna(axis=0)

print(f"Original shape: {df.shape}")
print(f"Numeric shape: {df_numeric.shape}")
print(f"Clean shape: {df_cl_baseline.shape}")

Original shape: (1460, 81)
Numeric shape: (1460, 38)
Clean shape: (1121, 38)


## Spliting the data, training and baseline resualts

In [7]:
from sklearn.model_selection import train_test_split, cross_val_score

In [8]:
X = df_cl_baseline.drop(['SalePrice', 'Id'], axis=1)
y = df_cl_baseline['SalePrice']

# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [9]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
cv_scores = cross_val_score(
    model,
    X,
    y,
    scoring="neg_root_mean_squared_error",
    cv=5
)
rmse_scores = -cv_scores

print("RMSE scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Std RMSE:", rmse_scores.std())


RMSE scores: [29587.07709133 37665.99103449 40100.65600937 32378.41339688
 54086.83339133]
Mean RMSE: 38763.79418468276
Std RMSE: 8517.900283988969


# Above is the result for our baseline model.

# Pipelines, Imputers, preprocessing and model

In [13]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import Ridge

In [14]:
X = df.drop(['SalePrice', 'Id'], axis=1)
y = df['SalePrice']

num_features = X.select_dtypes(include=['int64', 'float64']).columns

# for some features missing value has meaning like Garage year build (house has no garage), can't be replaced with median or mode.
# same with masanory area. so its best to fill with zero.
num_zero_features = ['GarageYrBlt', 'MasVnrArea']

# avoiding overlaping
num_features = num_features.drop(['GarageYrBlt', 'MasVnrArea'])


# cat_features = X.select_dtypes(include=['object']).columns

# categorical features with null value mean doesn't exit, i.e no fence, fireplace or garage
# best to replace with none
cat_none_features = ['PoolQC', 'MiscFeature', 'Alley', 'Fence', 'MasVnrType', 'FireplaceQu', 'GarageCond', 'GarageType', 'GarageFinish', 'GarageQual', 'BsmtFinType2', 'BsmtExposure',
             'BsmtQual', 'BsmtCond', 'BsmtFinType1']

# few missing values, best to be replaced with mode
cat_mode_features = ['Electrical']


In [39]:
num_median_transformer = Pipeline([(
    "imputer", SimpleImputer(strategy='median'))
])
num_zero_transformer = Pipeline([(
    "imputer", SimpleImputer(strategy='constant', fill_value=0))
])

cat_none_transformer = Pipeline([
    ("imputer", SimpleImputer(strategy="constant", fill_value="None")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])
cat_mode_transformer = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])


preprocessor = ColumnTransformer(
    transformers=[
        ("num_median", num_median_transformer, num_features),
        ("num_zero", num_zero_transformer, num_zero_features),
        ("cat_none", cat_none_transformer, cat_none_features),
        ("cat_mode", cat_mode_transformer, cat_mode_features)
    ]
)

pipeline = Pipeline([
    ('preprocess', preprocessor),
    ('model', Ridge(alpha=10))
])


# Finding the best alpha for ridge
manually
alpha=0.01   mean RMSE=36144 | std=8429  
alpha=0.1    mean RMSE=35334 | std=7744  
alpha=1.0    mean RMSE=34664 | std=7529  
alpha=10     mean RMSE=34151 | std=7583  
alpha=100    mean RMSE=34223 | std=7760  

**alpha = 10**
- Mean RMSE is clearly lower
- Std barely increases (still very stable)
- No sign of underfitting yet

In [19]:
alphas = [0.01, 0.1, 1.0, 10, 100]

for a in alphas:
    pipeline = Pipeline([
        ('preprocess', preprocessor),
        ('model', Ridge(alpha=a))
    ])
    
    cv_scores = cross_val_score(
        pipeline,
        X,
        y,
        scoring="neg_root_mean_squared_error",
        cv=5
    )
    
    rmse = -cv_scores
    print(f"alpha={a:<6} mean RMSE={rmse.mean():.0f} | std={rmse.std():.0f}")

alpha=0.01   mean RMSE=36144 | std=8429
alpha=0.1    mean RMSE=35334 | std=7744
alpha=1.0    mean RMSE=34664 | std=7529
alpha=10     mean RMSE=34151 | std=7583
alpha=100    mean RMSE=34223 | std=7760


**alpha = 10**
- Mean RMSE is clearly lower
- Std barely increases (still very stable)
- No sign of underfitting yet

# Running Ridge with CV and best alpha 

In [40]:
cv_scores = cross_val_score(
    pipeline,
    X,
    y,
    scoring="neg_root_mean_squared_error",
    cv=5
)

In [41]:
rmse_scores = -cv_scores

In [42]:
print("RMSE scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Std RMSE:", rmse_scores.std())

RMSE scores: [27170.41082736 35119.12815987 34024.61590893 26755.23139835
 47684.58121623]
Mean RMSE: 34150.79350214832
Std RMSE: 7583.364474980232


# Linear regression results
RMSE scores: [29599.84709821 35696.85172911 42926.73509051 24871.91900602
 48913.65432257]  
Mean RMSE: 36401.801449283426  
Std RMSE: 8704.285698468633  

# comparision results:  
after running data through a simple pipeline. the resaults are not that different since we only filled the numerical missing values. and there are not that many numerical missing values in our data.  
the missing values are about %95 categorical.  

around ~2k less error in mean RMSE

and that is using LinearRegression.

after using a more proper pipeline, with all the categorical and numerical features for Ridge model. the results are just a little bit better which is totally disapointing. but more realistic now.

# Extracting features and coefficients from Ridge model

In [21]:
pipeline.fit(X, y)

Pipeline(steps=[('preprocess',
                 ColumnTransformer(transformers=[('num_median',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median'))]),
                                                  Index(['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF',
       'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea',
       'BsmtFullBa...
                                                  ['PoolQC', 'MiscFeature',
                                                   'Alley', 'Fence',
                                                   'MasVnrType', 'FireplaceQu',
                                                   'GarageCond', 'GarageType',
                                                   'GarageFinish', 'GarageQual',
                                                   'BsmtFinType2',
                                                   'BsmtExposure', 'BsmtQual',
                                                   'BsmtCond',
                                                   'BsmtFinType1']),
                                                 ('cat_mode',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['Electrical'])])),
                ('model', Ridge(alpha=10))])

In [22]:
preprocessor = pipeline.named_steps['preprocess']

In [23]:
num_features_out = list(num_features) +  list(num_zero_features)

In [24]:
cat_none_ohe = preprocessor.named_transformers_["cat_none"] \
    .named_steps["onehot"] \
    .get_feature_names_out(cat_none_features)

cat_mode_ohe = preprocessor.named_transformers_["cat_mode"] \
    .named_steps["onehot"] \
    .get_feature_names_out(cat_mode_features)


In [25]:
feature_names = (
    num_features_out +
    list(cat_none_ohe) +
    list(cat_mode_ohe)
)

In [27]:
coefs = pipeline.named_steps["model"].coef_

In [29]:
# Sanity check
len(feature_names), len(coefs)

(120, 120)

In [30]:
coef_df = (
    pd.DataFrame({
        "feature": feature_names,
        "coefficient": coefs
    })
    .assign(abs_coef=lambda df: df["coefficient"].abs())
    .sort_values("abs_coef", ascending=False)
)


In [35]:
coef_df.head(20)

feature   coefficient      abs_coef
38            PoolQC_Gd -28986.121387  28986.121387
98          BsmtQual_Ex  25997.316081  25997.316081
36            PoolQC_Ex  21257.309805  21257.309805
3           OverallQual  13642.623450  13642.623450
23           GarageCars  13617.861048  13617.861048
20         KitchenAbvGr -12636.338925  12636.338925
57       FireplaceQu_Ex  12008.709221  12008.709221
94      BsmtExposure_Gd  11376.228383  11376.228383
80        GarageQual_Ex  10779.445414  10779.445414
100         BsmtQual_Gd -10336.825360  10336.825360
39          PoolQC_None   9262.534576   9262.534576
102         BsmtQual_TA  -8684.877972   8684.877972
53    MasVnrType_BrkCmn  -8259.552275   8259.552275
110    BsmtFinType1_GLQ   8174.271866   8174.271866
55      MasVnrType_None   7756.649474   7756.649474
96      BsmtExposure_No  -7240.302257   7240.302257
85        GarageQual_TA  -7070.885407   7070.885407
114    BsmtFinType1_Unf  -6918.429499   6918.429499
81        GarageQual_Fa  -6901.752479   6901.752479
72   GarageType_BuiltIn   6895.449646   6895.449646

In [37]:
coef_df.sort_values("coefficient", ascending=False).head(15)

feature   coefficient      abs_coef
98          BsmtQual_Ex  25997.316081  25997.316081
36            PoolQC_Ex  21257.309805  21257.309805
3           OverallQual  13642.623450  13642.623450
23           GarageCars  13617.861048  13617.861048
57       FireplaceQu_Ex  12008.709221  12008.709221
94      BsmtExposure_Gd  11376.228383  11376.228383
80        GarageQual_Ex  10779.445414  10779.445414
39          PoolQC_None   9262.534576   9262.534576
110    BsmtFinType1_GLQ   8174.271866   8174.271866
55      MasVnrType_None   7756.649474   7756.649474
72   GarageType_BuiltIn   6895.449646   6895.449646
4           OverallCond   6257.125103   6257.125103
15         BsmtFullBath   5833.021990   5833.021990
17             FullBath   5716.585188   5716.585188
22           Fireplaces   5592.794494   5592.794494

In [38]:
coef_df.sort_values("coefficient").head(15)

feature   coefficient      abs_coef
38           PoolQC_Gd -28986.121387  28986.121387
20        KitchenAbvGr -12636.338925  12636.338925
100        BsmtQual_Gd -10336.825360  10336.825360
102        BsmtQual_TA  -8684.877972   8684.877972
53   MasVnrType_BrkCmn  -8259.552275   8259.552275
96     BsmtExposure_No  -7240.302257   7240.302257
85       GarageQual_TA  -7070.885407   7070.885407
114   BsmtFinType1_Unf  -6918.429499   6918.429499
81       GarageQual_Fa  -6901.752479   6901.752479
99         BsmtQual_Fa  -5732.138462   5732.138462
19        BedroomAbvGr  -5392.689866   5392.689866
69   GarageType_2Types  -5163.189810   5163.189810
58      FireplaceQu_Fa  -4950.357405   4950.357405
62      FireplaceQu_TA  -4645.739421   4645.739421
111   BsmtFinType1_LwQ  -4249.416821   4249.416821

# Reports:
after a general look on the approach, decided that linear models are not suitable for this data at all, but it was all a good practice working with linear regression, cross validation and pipelines. I have a better understanding on these and also on our data as well.

I will be moving from linear models to other options
- Tree-based models
- XGBoost

tldr: thinking about only the performance of our model, 34k RMSE is absolutely awful, must try different approach.